pkgs/development/compilers/llvm/common/default.nix

```diff
@@ -1065,160 +1065,173 @@ let
+                    url = "https://github.com/llvm/llvm-project/commit/5e2358c781b85a18d1463fd924d2741d4ae5e42e.patch";
+                    hash = "sha256-UxIlAifXnexF/MaraPW0Ut6q+sf3e7y1fMdEv1q103A=";
                   }
-                )
-            ++
-              lib.optional
-                (
-                  lib.versionAtLeast metadata.release_version "18"
-                  && stdenv.hostPlatform.isDarwin
-                  && lib.versionOlder stdenv.hostPlatform.darwinMinVersion "10.13"
-                )
-                # https://github.com/llvm/llvm-project/issues/64226
-                (metadata.getVersionFile "libcxx/0001-darwin-10.12-mbstate_t-fix.patch");
-          stdenv =
-            if stdenv.hostPlatform.isDarwin then
-              overrideCC darwin.bootstrapStdenv buildLlvmTools.clangWithLibcAndBasicRt
-            else
-              overrideCC stdenv buildLlvmTools.clangWithLibcAndBasicRt;
-        }
-        // lib.optionalAttrs (lib.versionOlder metadata.release_version "14") {
-          # TODO: remove this, causes LLVM 13 packages rebuild.
-          inherit (metadata) monorepoSrc; # Preserve bug during #307211 refactor.
-        }
-      );
+                );
+        };
+      }
+      // lib.optionalAttrs (lib.versionAtLeast metadata.release_version "20") {
+        libc-overlay = callPackage ./libc {
+          isFullBuild = false;
+        };
 
-      libunwind = callPackage ./libunwind {
-        patches = lib.optional (lib.versionOlder metadata.release_version "17") (
-          metadata.getVersionFile "libunwind/gnu-install-dirs.patch"
-        );
-        stdenv = overrideCC stdenv buildLlvmTools.clangWithLibcAndBasicRt;
-      };
+        libc-full = callPackage ./libc {
+          isFullBuild = true;
+        };
 
-      openmp = callPackage ./openmp {
-        patches =
-          lib.optional (
-            lib.versionAtLeast metadata.release_version "15" && lib.versionOlder metadata.release_version "19"
-          ) (metadata.getVersionFile "openmp/fix-find-tool.patch")
-          ++ lib.optional (
-            lib.versionAtLeast metadata.release_version "14" && lib.versionOlder metadata.release_version "18"
-          ) (metadata.getVersionFile "openmp/gnu-install-dirs.patch")
-          ++ lib.optional (lib.versionAtLeast metadata.release_version "14") (
-            metadata.getVersionFile "openmp/run-lit-directly.patch"
-          )
-          ++
-            lib.optional (lib.versionOlder metadata.release_version "14")
-              # Fix cross.
-              (
-                fetchpatch {
-                  url = "https://github.com/llvm/llvm-project/commit/5e2358c781b85a18d1463fd924d2741d4ae5e42e.patch";
-                  hash = "sha256-UxIlAifXnexF/MaraPW0Ut6q+sf3e7y1fMdEv1q103A=";
-                }
-              );
-      };
-    }
+        libc = if stdenv.targetPlatform.libc == "llvm" then libraries.libc-full else libraries.libc-overlay;
```

Comment:
I think this attribute being for two different things depending on the platform is going to be confusing.

Reply:
Tbf, we already have attributes which confuse people. `clangUseLLVM`, `compiler-rt`, `stdenv`.

Reply:
And that's justification to make things more confusing??

Reply:
Well I think it's not that confusing, it'll cause some accidents until someone realizes which one they should use if they need it. The lib names are different between the two so someone could easily tell when things go wrong when an error happens. Also, documenting the attributes better would help a lot.

Reply:
So why not make it explicit whether full or overlay is being used?  As we've established, they're used completely differently, so there isn't a situation where you'd want an attribute that was sometimes the overlay and sometimes full.

Reply:
My thinking was having a shim which adds in both inside the overlay mode. Then the function which selects the libc cross could detect when to use the overlay mode. That could allow for mixing LLVM libc and glibc or LLVM libc and musl.

Reply:
From what I know, a shim can be made by writing a linker script into `libc.so`.

Reply:
Then how about we hold off on this attribute until that's ready to go?  It's not really useful as it is.

Reply:
I've added the shim.